File: dotnet/test/Microsoft.AutoGen.Core.Tests/HandlerInvokerTest.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// HandlerInvokerTest.cs

using FluentAssertions;
using Microsoft.AutoGen.Contracts;
using Xunit;

namespace Microsoft.AutoGen.Core.Tests;

[Trait("Category", "UnitV2")]
public class HandlerInvokerTest()
{
    public List<(string, MessageContext)> PublishlikeInvocations = new List<(string, MessageContext)>();

    public ValueTask PublishlikeAsync(string message, MessageContext messageContext)
    {
        this.PublishlikeInvocations.Add((message, messageContext));
        return ValueTask.CompletedTask;
    }

    public List<(string, MessageContext)> SendlikeInvocations = new List<(string, MessageContext)>();

    public ValueTask<int> SendlikeAsync(string message, MessageContext messageContext)
    {
        this.SendlikeInvocations.Add((message, messageContext));
        return ValueTask.FromResult(this.SendlikeInvocations.Count);
    }

    [Fact]
    public async Task Test_InvokingPublishlike_Succeeds()
    {
        MessageContext messageContext = new MessageContext(Guid.NewGuid().ToString(), CancellationToken.None);

        var methodInfo = typeof(HandlerInvokerTest).GetMethod(nameof(PublishlikeAsync))!;
        var invoker = new HandlerInvoker(methodInfo, this);

        object? result = await invoker.InvokeAsync("Hello, world!", messageContext);

        this.PublishlikeInvocations.Should().HaveCount(1);
        this.PublishlikeInvocations[0].Item1.Should().Be("Hello, world!");
        result.Should().BeNull();
    }

    [Fact]
    public async Task Test_InvokingSendlike_Succeeds()
    {
        MessageContext messageContext = new MessageContext(Guid.NewGuid().ToString(), CancellationToken.None);

        var methodInfo = typeof(HandlerInvokerTest).GetMethod(nameof(SendlikeAsync))!;
        var invoker = new HandlerInvoker(methodInfo, this);

        object? result = await invoker.InvokeAsync("Hello, world!", messageContext);

        this.SendlikeInvocations.Should().HaveCount(1);
        this.SendlikeInvocations[0].Item1.Should().Be("Hello, world!");
        result.Should().Be(1);
    }
}

File: typescript/test/core/HandlerInvoker.test.ts
import { describe, it, expect } from '@jest/globals';
import { HandlerInvoker } from '../../src/core/HandlerInvoker';
import { MessageContext } from '../../src/contracts/MessageContext';

class HandlerInvokerTest {
  public publishlikeInvocations: Array<[string, MessageContext]> = [];
  public sendlikeInvocations: Array<[string, MessageContext]> = [];

  async publishlikeAsync(message: string, messageContext: MessageContext): Promise<void> {
    this.publishlikeInvocations.push([message, messageContext]);
  }

  async sendlikeAsync(message: string, messageContext: MessageContext): Promise<number> {
    this.sendlikeInvocations.push([message, messageContext]);
    return this.sendlikeInvocations.length;
  }
}

describe('HandlerInvoker', () => {
  it('should successfully invoke publishlike method', async () => {
    const messageContext = new MessageContext(crypto.randomUUID());
    const testInstance = new HandlerInvokerTest();
    
    const handler = {
      handleAsync: testInstance.publishlikeAsync.bind(testInstance)
    };
    
    const invoker = new HandlerInvoker(handler);
    const result = await invoker.invoke("Hello, world!", messageContext);

    expect(testInstance.publishlikeInvocations).toHaveLength(1);
    expect(testInstance.publishlikeInvocations[0][0]).toBe("Hello, world!");
    expect(result).toBeUndefined();
  });

  it('should successfully invoke sendlike method', async () => {
    const messageContext = new MessageContext(crypto.randomUUID());
    const testInstance = new HandlerInvokerTest();
    
    const handler = {
      handleAsync: testInstance.sendlikeAsync.bind(testInstance)
    };
    
    const invoker = new HandlerInvoker(handler);
    const result = await invoker.invoke("Hello, world!", messageContext);

    expect(testInstance.sendlikeInvocations).toHaveLength(1);
    expect(testInstance.sendlikeInvocations[0][0]).toBe("Hello, world!");
    expect(result).toBe(1);
  });
});
